### ISSUE TITLE
Implement Polygon.Intersection

### ISSUE THREAD

**Issue by ctessum** [golang/geo#44]
Since it seems acceptable to use this space to indicate interest in specific features, I would like to request polygon intersections. The Go language currently lacks a high quality library for boolean operations on geographic features, and having one would help me out greatly. (I am one of the maintainers of a low-to-medium quality library.) I greatly appreciate all the work that has happened so far to port this library to Go.

**Comment by dsymonds** [comment]
Yep, it is totally acceptable.  ;-)

e7cfa0c added Polygon.Contains and Polygon.Intersects. To clarify, you're after the equivalent of C++'s `S2Polygon::InitToIntersection`?

**Comment by ctessum** [comment]
Yes, I mean the equivalent of `S2Polygon::InitToIntersection`. Thanks!

**Comment by brendanashworth** [comment]
I'm interested in contributing a PR for this — are there any thoughts on using [Martinez-Rueda](http://www.cs.ucr.edu/~vbz/cs230papers/martinez_boolean.pdf) for the boolean operation algorithm? Or would it be best to port the C++ version accordingly?

**Comment by ctessum** [comment]
There is a port of (one of) the Martinez-Rueda algorithms [here](https://github.com/ctessum/polyclip-go). However, that one is for planar geometry, not sure that it would work here?

**Comment by brendanashworth** [comment]
@ctessum I'm not sure myself. The projection onto a plane would conserve intersection points but not lengths.. is it naive to assume the lengths are proportional and thus area is conserved after we project back to the sphere?

**Comment by alphairys** [comment]
any update on this? Curious if anyone is actively working on this piece?

**Comment by brendanashworth** [comment]
@alphairys I don't have an update/I haven't been working on it. If you're looking for a workaround, I've been using CellUnions (after [covering the polygon with cells](https://godoc.org/github.com/golang/geo/s2#RegionCoverer)) as a workaround, and they implement simple binary operations like [CellUnionFromIntersection](https://godoc.org/github.com/golang/geo/s2#CellUnionFromIntersection).

**Comment by rsned** [comment]
There is still work being done on this

**Comment by ipipdotnet** [comment]
I think someone has already done that  https://github.com/davidreynolds/gos2/blob/master/s2/polygon_test.go

**Comment by panmari** [comment]
It looks like the repository at https://github.com/davidreynolds/gos2/blob/master/s2/polygon_test.go is not active anymore. Is anyone interested in porting the code to golang/geo?
